### PR TITLE
WIP bpo-34589: Move locale coercion back to the CLI app

### PR DESCRIPTION
**Pull request by ncoghlan**
This PR moves locale coercion back to happening as soon as possible in the CLI app, so there's no re-encoding dance involved in handling it, and so you can emulate systems that don't have a suitable target locale with `PYTHONCOERCECLOCALE=0`, even when working on code that runs with `-E` or `-I`.

To help with the documentation enhancements in the PR, it also implements [[bpo-30672](https://bugs.python.org/issue30672)](https://www.bugs.python.org/issue30672) (which makes the POSIX locale behave the same way as it does on glibc, even when it's not a simple alias for the default C locale)

https://bugs.python.org/issue34589

### CODE REVIEW

File: Python/pylifecycle.c
Comment (ncoghlan) [inline Python/pylifecycle.c]: This function is missing checks for `coercion_target` and `coercion_warning` being NULL, and skipping the assignments in that case.

**Review by vstinner — DISMISSED** [review]
I dislike the idea of having PYTHONCOERCECLOCALE env var which behaves differently than all other PYTHON* env vars. See my -X proposal: https://bugs.python.org/issue34589#msg325282

**Comment by vstinner** [comment]
I merged my PR #9073 instead: it is closer to the current code, allows to implement a new -X coerce_c_locale command line option, and still respects -E and -I option: ignore PYTHONCOERCECLOCALE.

Maybe we can refactor the Python initialization code later, but we need a fix on the current code before 3.7.1 release which is scheduled tomorrow by our release manager, Ned.

**Comment by ncoghlan** [comment]
What the hell Victor? This is NOT OK!

**Comment by ncoghlan** [comment]
I've dismissed Victor's review, as he is inappropriately attempting to overrule the PEP process due to his personal preferences.

**Comment by ncoghlan** [comment]
@vstinner "Closer to the current code" is not an advantage when the structure of the current code goes completely against the way PEP 538 was designed to work (i.e. the CPython runtime never even seeing the original uncoerced locale, except in the locale coercion helper functions)

**Comment by ncoghlan** [comment]
Initial Travis failure affecting https://github.com/python/cpython/pull/9257/commits/bb16468bb69fb721ceadfd89f262d6b3172d4d6b was in test_nntplib relating to being unable to assign an address, so I restarted the job.

File: Python/pylifecycle.c
Comment (ncoghlan) [reply Python/pylifecycle.c]: I added these missing checks in https://github.com/python/cpython/pull/9257/commits/f4c686babca9870fc832797f2b960a6420c14f38

**Comment by ncoghlan** [comment]
Got the test_nntplib failure again, so restarted all the CI checks. See https://bugs.python.org/issue19756 (in particular, the last couple of comments)

**Comment by ncoghlan** [comment]
@vstinner I know you object-in-principle to the idea of having `PYTHONCOERCECLOCALE` behave the same way as `LANG`, `LC_CTYPE`, and `LC_ALL` do (rather than having it behave like other `PYTHON*` settings), but I'd still appreciate it if you were willing to review the changes I'm proposing to the CoreConfig initialisation code.

The big items are:

- the coercion related settings go away, replaced with a simple setting related solely to displaying C locale related warnings (or not)
- a separate `read_unconditional_env_vars` helper function is added

(This isn't urgent, since it's currently only for Python 3.8, and even if a variant of it does get backported, it would be for 3.7.2 at the earliest)

**Comment by vstinner** [comment]
Sorry, I will not review this change. (If you want my opinion, I dislike the PR since it introduces an exception to the -E option and I dislike that.)

File: Modules/main.c
Comment (ericsnowcurrently) [inline Modules/main.c]: accidental blank line?

File: Modules/main.c
Comment (ericsnowcurrently) [inline Modules/main.c]: mis-indented?

File: Python/pylifecycle.c
Comment (ericsnowcurrently) [inline Python/pylifecycle.c]: Did you mean for one of these two to be `"c"` (lower-case) or something?

File: Modules/main.c
Comment (ericsnowcurrently) [inline Modules/main.c]: The function comment above for `pymain_read_conf()` is out-of-date with this change, no?

File: Python/coreconfig.c
Comment (ericsnowcurrently) [inline Python/coreconfig.c]: wrong indentation?

File: Python/pylifecycle.c
Comment (ericsnowcurrently) [inline Python/pylifecycle.c:421]: It may be worth noting somewhere that this is one of the few C-API functions that may be called before the runtime is configured (or possibly even initialized).  I know it's somewhat orthogonal to the objectives of this PR, but such a comment would give a little context.

File: Modules/main.c
Comment (ericsnowcurrently) [inline Modules/main.c:1794]: If we used sprintf in `_Py_CoerceLegacyLocale()` then we'd only need to return the fully-formed warning from it and we could drop `pymain->locale_coercion_target` entirely.  I mention this because that field was a little misleading/confusing at first (I was expecting it be used elsewhere).

Did you have some other plan for that field?

**Comment by ericsnowcurrently** [comment]
Sorry this is taking so long.  I wasn't super familiar with the locale stuff so I've been taking my time to make sure I've got it straight.  Let me make sure I understand the change correctly.

Status quo:

1. deal with commandline/env (and locale)
    1. init runtime state
    2. default to `Py_UTF8Mode`
    3. call `_Py_SetLocaleFromEnv(LC_CTYPE)`
    4. parse commandline
    5. maybe populate core config from env vars (with locale already set), including `coerce_c_locale*` (from `PYTHONCOERCECLOCALE`)
    6. call `_Py_CoerceLegacyLocale()` if `coerce_c_locale` is set (from step 1.5)...and go back to step 1.4
    7. go back to step 1.4 if switched to UTF8 mode
2. init stdio
3. init core
4. configure main interpreter
5. initialize main interpreter
6. run the program

This PR:

1. call `_Py_SetLocaleFromEnv(LC_CTYPE)`
2. call `Py_CoerceLegacyLocale()` if a legacy locale is detected *and* `PYTHONCOERCECLOCALE` is set
3. deal with commandline/env (and locale)
    1. init runtime state
    2. default to `Py_UTF8Mode`
    3. parse commandline
    4. maybe populate core config from env vars, *not* including `coerce_c_locale*`
    5. always populate `warn_c_locale` (using the locale-encoded value from `PYTHONCOERCECLOCALE`)
    6. go back to step 3 if switched to UTF8 mode
4. init stdio
5. print the locale coercion warning if appropriate
6. init core
7. configure main interpreter
8. initialize main interpreter
9. run the program

Here are the differences I noticed:

Status quo:
* 2 `_PyCoreConfig` fields (one to indicate that coercion should happen & one for the warning)
* initial locale setting (with `LC_CTYPE`)
    * right before processing commandline (step 1.3 above)
* reading `PYTHONCOERCECLOCALE` (step 1.5 above)
    * not used if `-E` was used
    * relies on currently-set locale (possibly UTF mode)
* `_Py_LegacyLocaleDetected()`
    * called in `config_init_locale()` (basically right after env vars are processed; step 1.5 above)
    * used as a fallback if `PYTHONCOERCECLOCALE` wasn't set
* `_Py_CoerceLegacyLocale()`
    * called when in the midst of processing the commandline (step 1.6 above)
    * called if `PYTHONCOERCECLOCALE` set **or** `_Py_LegacyLocaleDetected()` returns true
* warning
    * happens immediately when `_Py_CoerceLegacyLocale()` is called (step 1.6 above)

This PR:
* 1 `_PyCoreConfig` field (to indicate the warning)
* the locale is initialized to `LC_CTYPE` immediately (step 1 above)
* reading `PYTHONCOERCECLOCALE` (step 2 above)
    * used even if `-E` was used
    * relies only on `LC_CTYPE` locale
* reading `PYTHONCOERCECLOCALE` (step 3.5 above)
    * only used if `-E` was used
    * relies on currently-set locale
* `_Py_LegacyLocaleDetected()`
    * called right away, before any other locale stuff might have been done or any env vars/cmdline processed (step 2 above)
* `_Py_CoerceLegacyLocale()`
    * also called right away and only if a legacy locale is detected (step 2 above)
    * called if `PYTHONCOERCECLOCALE` set (and not "0") **and** `_Py_LegacyLocaleDetected()` returns true
* warning
    * happens right before core initialization (step 5 above)

In practice, it seems that the key changes are:
* `_Py_CoerceLegacyLocale()` is only called if a legacy locale is detected *and* `PYTHONCOERCECLOCALE` is set
* `-E` no longer affects the decision to call `_Py_CoerceLegacyLocale()`

File: Misc/NEWS.d/next/Core and Builtins/2018-09-20-21-36-57.bpo-34589.9zERCD.rst
Comment (ericsnowcurrently) [inline Misc/NEWS.d/next/Core and Builtins/2018-09-20-21-36-57.bpo-34589.9zERCD.rst:3]: I'm not sure I follow the distinction you are describing here.  As far as I can tell, the role of the embedding app has not changed at all.  By "CPython runtime" are you talking about the internal C-API?  Otherwise I'm just not seeing it.

Furthermore, from what I can tell, this change is really about 2 specific things:
* make `PYTHONCOERCECLOCALE` independent of the `-E` flag
* only use `PYTHONCOERCECLOCALE` if a legacy locale is detected

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
The change makes sense to me, especially in light of the following from PEP 538:

> The interpreter will always check for the PYTHONCOERCECLOCALE environment variable at startup (even when running under the -E or -I switches), as the locale coercion check necessarily takes place before any command line argument processing. For consistency, the runtime check to determine whether or not to suppress the locale compatibility warning will be similarly independent of these settings.

There are a few minor things to address otherwise.

One point of clarification:  is it intentional that the warning is not emitted under `-E` even if the locale is coerced?

File: Modules/main.c
Comment (ncoghlan) [reply Modules/main.c]: Deleted

File: Modules/main.c
Comment (ncoghlan) [reply Modules/main.c]: Yep

File: Python/pylifecycle.c
Comment (ncoghlan) [reply Python/pylifecycle.c]: Copy-and-paste error (second one was supposed to POSIX), combined with deferral of the related tests to a follow-up PR for https://www.bugs.python.org/issue30672

File: Python/pylifecycle.c
Comment (ncoghlan) [reply Python/pylifecycle.c:421]: Added a note to both this and the legacy locale detection API.

The downside of the notes I added is that a lot of other functions have those same restrictions, but *don't* have a marker comment like the ones I just added, so they may give folks an incorrect impression.

So it may be better to take those comments out again, and file an enhancement issue on BPO pointing out that this is a maintainability issue, where it's not obvious which parts of the code need to be careful about only calling pre-init friendly APIs.

File: Modules/main.c
Comment (ncoghlan) [reply Modules/main.c]: Clarified (also filed https://bugs.python.org/issue34914 which I was planning to do anyway, updating that comment just reminded me I hadn't done it yet)

File: Python/coreconfig.c
Comment (ncoghlan) [reply Python/coreconfig.c]: Fixed.

File: Modules/main.c
Comment (ncoghlan) [reply Modules/main.c:1794]: The issue I had with doing the message formatting in _Py_CoerceLegacyLocale is coping with the dynamic memory allocation needed (you could use a statically allocated array instead, but that's a different kind of ick).

However, I'm thinking there may be a better way to defer the rendering of the warning while still only needing one field on the main struct: make the field a function pointer accepting a C output stream that gets called instead of a C string that gets printed, and then in the warning rendering function, substitute the actual current locale into the message, rather than saving the name of the coercion target (This is all internal to the CPython command line app, so we don't need to worry about intervening locale changes).

With that change, this would just be a standard C callback API, rather than something that tightly couples the code that determines the warning is needed to the code that actually prints it after the standard streams are configured.

File: Misc/NEWS.d/next/Core and Builtins/2018-09-20-21-36-57.bpo-34589.9zERCD.rst
Comment (ncoghlan) [reply Misc/NEWS.d/next/Core and Builtins/2018-09-20-21-36-57.bpo-34589.9zERCD.rst:3]: As released in 3.7.0, locale coercion happens at the point where `Py_Initialize` gets called (as that then internally coerces the locale to something else). That's appropriate for UTF-8 mode (since that only affects the CPython runtime in the current process, and the code carefully goes back and recalculates the settings that were already determined based on the locale encoding), but not for locale coercion (since that affects all subsequently executed locale-aware code in the current process, as well as in subprocesses, but can't reliably go back and rewrite all previously determined values in the embedding application).

That's not the way PEP 538 was originally implemented, and deliberately so: instead, the locale coercion happened in *the CLI app*, before the first call to `Py_DecodeLocale`, so nothing would ever get decoded as ASCII in the first place.

While that CLI app code has since moved to a private API inside the shared library (as `_Py_UnixMain`), this PR still gets it back to the same relative location: before the first call to `Py_DecodeLocale`, so its effectively indistinguishable from `LC_CTYPE=C.UTF-8` actually being set externally in the environment.

File: Python/coreconfig.c
Comment (ncoghlan) [inline Python/coreconfig.c]: Note to self: still need to revert this.

File: Python/pylifecycle.c
Comment (ncoghlan) [inline Python/pylifecycle.c]: I was hoping to get away without needing to rely on the full `getopt` argument definition string here, but realised that's a futile endeavour due to the fact that cases like `python -iEV` are perfectly valid option strings.

File: Misc/NEWS.d/next/Core and Builtins/2018-09-20-21-36-57.bpo-34589.9zERCD.rst
Comment (ncoghlan) [reply Misc/NEWS.d/next/Core and Builtins/2018-09-20-21-36-57.bpo-34589.9zERCD.rst:3]: A potential analogy for why the way 3.7 is handling this is broken (and why PEP 538 deliberately defined it as working differently): performing locale coercion in `Py_Main` and `Py_Initialize` is akin to a Python library reconfiguring the logging module as a side effect of doing something else.

File: Python/pylifecycle.c
Comment (ncoghlan) [reply Python/pylifecycle.c]: Not so futile after all - with all the cross-platform getopt code migrated to `wchar_t`, it proved simpler to keep this early pass looking specifically for `-E` and `-I` independent of it.

Even adding a `-X no_locale_coercion` option would be straightforward (but I've chosen not to implement that at this stage)

File: Misc/NEWS.d/next/Core and Builtins/2018-09-20-21-36-57.bpo-34589.9zERCD.rst
Comment (ncoghlan) [reply Misc/NEWS.d/next/Core and Builtins/2018-09-20-21-36-57.bpo-34589.9zERCD.rst:3]: Amusingly enough, from the comments in the `readline` module implementation:

```
/* GNU readline() mistakenly sets the LC_CTYPE locale.
 * This is evil.  Only the user or the app's main() should do this!
 * We must save and restore the locale around the rl_initialize() call.
 */
```

**Comment by ncoghlan** [comment]
While messing with the current locale setting from a support library rather than the main application still bothers me, I've come to the conclusion that it doesn't bother me *enough* to want to introduce a non-trivial distinction between the way Python 3.7 handles this scenario, and the way Python 3.8+ handles it.

https://bugs.python.org/issue34589#msg332123 provides a bit more rationale (essentially, I expect that the the combination of circumstances required for there to be an externally observable difference in behaviour between Python 3.7.0 and this PR is going to be sufficiently rare that it simply isn't worth adding a net 200+ lines of code to maintain).

Accordingly, closing this PR and the associated issue indefinitely - while I preferred the original simpler implementation of locale coercion in isolation, all the machinery that the current more complex implementation relies on has to exist to handle UTF-8 mode anyway, so reverting locale coercion back to its original implementation model would end up making the interpreter as a whole more complicated (both to maintain, and to understand).

**Pull request closed without merging**
